flink-runtime/src/main/java/org/apache/flink/runtime/state/metrics/AbstractLatencyTrackingStateMetric.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.metrics;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.metrics.Histogram;
+import org.apache.flink.metrics.HistogramStatistics;
+import org.apache.flink.metrics.MetricGroup;
+
+import com.codahale.metrics.SlidingTimeWindowReservoir;
+import com.codahale.metrics.Snapshot;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/** Abstract class of latency tracking state metric which counts and histogram the state metric. */
+class AbstractLatencyTrackingStateMetric implements AutoCloseable {
+    protected static final String STATE_CLEAR_LATENCY = "stateClearLatency";
+    private final MetricGroup metricGroup;
+    private final int sampleInterval;
+    private final Map<String, Histogram> histogramMetrics;
+    private final Map<String, Counter> countersPerMetric;
+    private final Supplier<com.codahale.metrics.Histogram> histogramSupplier;
+
+    AbstractLatencyTrackingStateMetric(
+            String stateName, MetricGroup metricGroup, int sampleInterval, long slidingWindow) {
+        this.metricGroup = metricGroup.addGroup(stateName);
```

Comment:
How does the metric name look like in the end?
Can we distinguish between different subtasks if they are running in the same TM?

Reply:
I'll run a job to give more details.

Reply:
I used the `TopSpeedWindowing` example with log4j metrics reporter and get histograms below:

~~~
-- Histograms -----------------------------------------------------------------
host.taskmanager.container_id.CarTopSpeedWindowingExample.Window(GlobalWindows(), DeltaTrigger, TimeEvictor, ComparableAggregator, PassThr.0.last-element.valueStateGetLatency: count=8, min=915, max=32867, mean=10403.25, stddev=13057.645836279766, p50=1624.0, p75=22240.75, p95=32867.0, p98=32867.0, p99=32867.0, p999=32867.0
host.taskmanager.container_id.CarTopSpeedWindowingExample.Window(GlobalWindows(), DeltaTrigger, TimeEvictor, ComparableAggregator, PassThr.0.last-element.valueStateUpdateLatency: count=1, min=15139, max=15139, mean=15139.0, stddev=0.0, p50=15139.0, p75=15139.0, p95=15139.0, p98=15139.0, p99=15139.0, p999=15139.0
host.taskmanager.container_id.CarTopSpeedWindowingExample.Window(GlobalWindows(), DeltaTrigger, TimeEvictor, ComparableAggregator, PassThr.0.window-contents.listStateGetLatency: count=1, min=32406, max=32406, mean=32406.0, stddev=0.0, p50=32406.0, p75=32406.0, p95=32406.0, p98=32406.0, p99=32406.0, p999=32406.0
host.taskmanager.container_id.CarTopSpeedWindowingExample.Window(GlobalWindows(), DeltaTrigger, TimeEvictor, ComparableAggregator, PassThr.0.window-contents.stateClearLatency: count=1, min=34417, max=34417, mean=34417.0, stddev=0.0, p50=34417.0, p75=34417.0, p95=34417.0, p98=34417.0, p99=34417.0, p999=34417.0
host.taskmanager.container_id.CarTopSpeedWindowingExample.Window(GlobalWindows(), DeltaTrigger, TimeEvictor, ComparableAggregator, PassThr.0.window-contents.listStateAddLatency: count=47, min=202, max=108342, mean=4574.638297872339, stddev=16627.402990016646, p50=311.0, p75=473.0, p95=28880.599999999882, p98=108342.0, p99=108342.0, p999=108342.0

~~~

Reply:
I guess `0` in `PassThr.0.window-contents.listStateAddLatency` is the subtask index - then should be fine.
Thanks!